File: Example/Example_3/example_3.c
// This is part of a contrived example to demonstrate
// the use of interoperable global constants. The constants
// are declared in the header, initialized in a void function
// here, and printed later by the Fortran program.

#include "example_3.h"
#include <string.h>

// Initialize all the variables declared in the header.
void initialize_vars() {
  int i = 0;
  int j = 0;
  for (i = 0; i < 4; i++) {
    for (j = 0; j < 3; j++) {
      our_global_numbers[i][j] = i * j - 2;
    }
  }

  // Note that initializing this as a string literal, ie
  // my_global_char = "MYLITERAL" will not work. Fortran
  // does not properly access C static or dynamic memory.
  strncpy(my_global_char, "MYLITERAL", 9);

  my_global_double = 4.5;
  _my_global_other = 0;
} 

File: Example/Example_3/example_3.f90
! This is the Fortran driver program which 
! calls the C defined initialization routine
! and then prints out the values of the variables
! as defined in C.

PROGRAM example_3
USE, INTRINSIC :: iso_c_binding
USE module_example_3
implicit none

INTEGER i, j ! Loop counters

h2m_my_global_other = 9 ! This is set to zero in C

! Initialize the variables in C.
CALL initialize_vars

WRITE(*,*) "Our character is: "
WRITE(*,*) my_global_char
WRITE(*,*) "Our global double is: "
WRITE(*,*) my_global_double
WRITE(*,*) "Our _my_global_other/h2m_my_global_other is: "
WRITE(*,*) h2m_my_global_other

! Print out the small array (which is twelve numbers long)
WRITE(*,*) "Our array is: "
do i = 1, 3
  do j = 1, 4
    WRITE(*,*) our_global_numbers(j, i)
  end do
end do

END PROGRAM example_3
